### PR TITLE
java interface support

### LINKED ISSUE #100: [Java Generator] Generate interface instead of abstract class for Java
Generate interface instead of abstract class for Java

This is a featured, but not an emergency. I will put it here only to we know that it can be done and how.

Source:
https://github.com/Asmot/djinni-rn/commit/84ebea92aa0fed8a2d28c4c7305463c17a7a9128

### PR DESCRIPTION
**Pull request by paulocoutinhox**
It check java files to be generated and if have only methods to be implemented, change final file to an `interface` instead of `abstract class`.

We can also add a flag for this if you want. What you think @LiFengSC?

closes #100

### CODE REVIEW

**Comment by LiFengSC** [comment]
Thanks for the PR.  This is a breaking change (meaning user code needs to be fixed to work with the new codegen). So I think we need to make this optional.

**Comment by paulocoutinhox** [comment]
@LiFengSC a flag, correct? Do you have a name?

**Comment by LiFengSC** [comment]
> @LiFengSC a flag, correct? Do you have a name?

"java-gen-interface" ?

this is similar to the existing "objc-gen-protocol"

btw,  with objc protocols,  we still generate protocols even for interfaces with statics,  but we put statics in a different place.

Another thing you probably need to pay attention to is importing YAML types.  If something is generated as interfaces but later imported to another project that are not using the flag, then there will be a mismatch.  You can look up the implementation of "objc-gen-protocol" to see how this was done.

File: src/source/JavaGenerator.scala [hunk 1]
```diff
@@ -174,8 +174,20 @@ class JavaGenerator(spec: Spec) extends Generator(spec) {
...
-      w.w(s"${javaClassAccessModifierString}abstract class $javaClass$typeParamList").braced {
+
+      // if no static and no cpp will use interface instead of abstract class
+      var classOrInterfaceDesc = "abstract class";
+      var methodPrefixDesc = "public abstract ";
+
+      if (!statics.nonEmpty && !i.ext.cpp) {
+        classOrInterfaceDesc = "interface"
+        methodPrefixDesc = ""
+      }
```
[the remark this answers is not in the document]
Comment (paulocoutinhox) [reply src/source/JavaGenerator.scala]: done

**Comment by paulocoutinhox** [comment]
> > @LiFengSC a flag, correct? Do you have a name?
> 
> "java-gen-interface" ?
> 
> this is similar to the existing "objc-gen-protocol"
> 
> btw, with objc protocols, we still generate protocols even for interfaces with statics, but we put statics in a different place.
> 
> Another thing you probably need to pay attention to is importing YAML types. If something is generated as interfaces but later imported to another project that are not using the flag, then there will be a mismatch. You can look up the implementation of "objc-gen-protocol" to see how this was done.

**About the Generator:**

DONE

**About the YAML:**

The only place that objc use the yaml info about protocol is here:

```
case e: MExtern => e.body match {
  case i: Interface =>
    if(i.ext.objc || e.objc.protocol) // <------- HERE
      (s"id<${e.objc.typename}>", false)
    else
      (e.objc.typename, true)
  case _ =>
    if (e.objc.generic) (e.objc.typename + args(tm), e.objc.pointer)
    else if(needRef) (e.objc.boxed, true)
    else (e.objc.typename, e.objc.pointer)
}
```

And java only use this for external:

```
case e: MExtern => (if(needRef) e.java.boxed else e.java.typename) + (if(e.java.generic) args(tm) else "")
```

I think that don't make sense for java. What do you think?

**Comment by paulocoutinhox** [comment]
I made a project to test generating and importing yaml files with it:
https://github.com/paulocoutinhox/djinni-java-interface-tests

**Comment by LiFengSC** [comment]
> > > @LiFengSC a flag, correct? Do you have a name?
> > 
> > 
> > "java-gen-interface" ?
> > this is similar to the existing "objc-gen-protocol"
> > btw, with objc protocols, we still generate protocols even for interfaces with statics, but we put statics in a different place.
> > Another thing you probably need to pay attention to is importing YAML types. If something is generated as interfaces but later imported to another project that are not using the flag, then there will be a mismatch. You can look up the implementation of "objc-gen-protocol" to see how this was done.
> 
> **About the Generator:**
> 
> DONE
> 
> **About the YAML:**
> 
> The only place that objc use the yaml info about protocol is here:
> 
> ```
> case e: MExtern => e.body match {
>   case i: Interface =>
>     if(i.ext.objc || e.objc.protocol) // <------- HERE
>       (s"id<${e.objc.typename}>", false)
>     else
>       (e.objc.typename, true)
>   case _ =>
>     if (e.objc.generic) (e.objc.typename + args(tm), e.objc.pointer)
>     else if(needRef) (e.objc.boxed, true)
>     else (e.objc.typename, e.objc.pointer)
> }
> ```
> 
> And java only use this for external:
> 
> ```
> case e: MExtern => (if(needRef) e.java.boxed else e.java.typename) + (if(e.java.generic) args(tm) else "")
> ```
> 
> I think that don't make sense for java. What do you think?

You are right,  Java does not need special syntax for interface vs class

**Review by LiFengSC — APPROVED** [review]
Looks good to me.  Thank you for spending your time on the PR and testing. 

FYI @bholmes-sc I'm approving this PR because this is something requested by the community.  It's behind a switch that defaults to false so it will not impact our use case.

**Pull request merged**